proposals/handler.go

```diff
@@ -356,19 +353,17 @@ func (h *Handler) handleProposal(ctx context.Context, expHash types.Hash32, peer
 		return err
 	}
 	proposalDuration.WithLabelValues(fetchTXs).Observe(float64(time.Since(t4)))
-
 	logger.With().Debug("proposal is syntactically valid")
-	t5 := time.Now()
-	if err := proposals.Add(h.db, &p); err != nil {
-		if errors.Is(err, sql.ErrObjectExists) {
-			known.Inc()
-			return fmt.Errorf("%w proposal %s", errKnownProposal, p.ID())
-		}
-		logger.With().Error("failed to save proposal", log.Err(err))
-		return fmt.Errorf("save proposal: %w", err)
+
+	err = h.proposals.OnProposal(&p)
+	switch {
+	case errors.Is(err, store.ErrProposalExists):
+		known.Inc()
+		return fmt.Errorf("%w %s", errKnownProposal, p.ID())
+	case err != nil:
+		return fmt.Errorf("saving proposal: %w", err)
```

Comment:
could you please doublecheck that in both of this cases p2p handler will return Ignore and not Reject. Reject will disconnect a peer, and we don't want that if proposal was evicted or known

Reply:
Perhaps we should refactor (in a separate effort) the handlers a bit to not use errors for control flow? It could return 2 things:
- a "decision" like `Ok|Ignore|Reject`,
- the error

Reply:
I checked - the p2p `Reject` is returned only if the error is `pubsub.ErrValidationReject`. Any other error results in `Ignore`. https://github.com/spacemeshos/go-spacemesh/blob/61cf5b72667bbf4381e5bb25bba66cf9ef5e49cd/p2p/pubsub/wrapper.go#L44-L51.